### PR TITLE
More stable algorithm to calculate zeros of spherical hankel function

### PR DESCRIPTION
**Pull request by fietew**
Follow up on #71: 

Based on the [publication](http://www.aes.org/e-lib/browse.cfm?elib=18609) of @narahahn and @spors.

This is basically a port of a [scipy routine](https://github.com/scipy/scipy/blob/v0.19.1/scipy/signal/filter_design.py#L3666-L3702).

@narahahn: Could you try out some parametrisation and check, if it is working as expected?

### CODE REVIEW

**Comment by fietew** [comment]
@hagenw: Is it okay to remove the other methods?

**Comment by hagenw** [comment]
The old version of the function returned `z` and `p` both as column vectors. Your new implementation returns `z` as row vector and `p` as column vector. Is this desired or a bug?

If the new method works better than the old ones, I see no big deal in replacing them as all our methods we implemented so far were buggy.

**Comment by hagenw** [comment]
Is the code ready for me to have a look at it?

It would be cool if you could add a test function for this, maybe doing something like the figures in #57 for example [those two](https://github.com/sfstoolbox/sfs-matlab/issues/57#issuecomment-186291311) or Figure 3 in @narahahn [publication](http://www.aes.org/e-lib/browse.cfm?elib=18609)

**Comment by fietew** [comment]
It is ready. I will add a test function.

**Comment by hagenw** [comment]
The test function seems to work ok under Octave, but I had to disable the legend as it covered the whole figure. I guess you have split it to six different subplots in order to not have to many overlapping points, or are the subplots showing different things?
For me they looked very similar, here is the first one of them, I guess this is the desired result?
sphbessel_zeros

**Comment by hagenw** [comment]
I was looking to update the reference as the Campos and Calderón paper was published in 2012, see http://doi.org/10.1155/2012/873078
The problem is that they reformulated some of their equations and I was not able to fit them easily to the ones in https://arxiv.org/pdf/1105.0957.pdf, which should be the paper you referenced. Could you have a look at it and propose if we should stay with the arxiv version or switch to the other one?
I would then do the update to the comments.

I'm also wondering if we should add a link to the [scipy routine](https://github.com/scipy/scipy/blob/v0.19.1/scipy/signal/filter_design.py#L3666-L3702) as you used it as a template.

Otherwise, I'm really happy that this seems to be finally solved as it bothered me since the first version that included NFC-HOA.

**Comment by fietew** [comment]
> I was looking to update the reference as the Campos and Calderón paper was published in 2012, see http://doi.org/10.1155/2012/873078
The problem is that they reformulated some of their equations and I was not able to fit them easily to the ones in https://arxiv.org/pdf/1105.0957.pdf, which should be the paper you referenced. Could you have a look at it and propose if we should stay with the arxiv version or switch to the other one?
I would then do the update to the comments.

Yes, I also had a look at the journal paper, which covers a more generalised problem and is harder to understand. I would stay with arxiv version since the formulas in the ``campos_zeros`` correspond directly to equations in the paper.

> I'm also wondering if we should add a link to the scipy routine as you used it as a template.

Yes, that's a good idea.

**Comment by hagenw** [comment]
Ok, I updated the references accordingly.

For the test function, could you maybe send me a screen shot of the Matlab result, that I can try to get the same with Octave.

File: SFS_general/sphbesselh_zeros.m [hunk 1]
```diff
@@ -53,74 +66,114 @@
...
+    return
+end
+
+% Approximate roots of nth-order ordinary Bessel polynomial as starting points
+r = campos_zeros(order);
+
+% Zeros of nth-order ordinary Bessel polynomial are the same as for
+% exp(1/x)*K_{n+0.5}(1/x), where K_v is the modified Bessel function of second
+% kind. We, hence, define the target function and its first derivative as
+f = @(x) besselk(order+0.5, 1./x, 1);  % modified Bessel function scaled by exp
```
Comment (hagenw) [inline SFS_general/sphbesselh_zeros.m]: I would remove the comment `% modified Bessel function scaled by exp` as it my lead a user to expect the `exp()` function to be part of that line. In addition, I think the comment provided before is sufficient.

Comment (fietew) [inline SFS_general/sphbesselh_zeros.m]: It took me a while, to figure out that ``besselk(nu, x, 1)`` is equal to ``besselk(nu,x)*exp(x)``. This is why I added this comment. But if it misleading, we can remove it.

Comment (hagenw) [inline SFS_general/sphbesselh_zeros.m]: I was a little bit confused by it.
Maybe it helps if we state explicitly the equation in the first comment. Something like
```Matlab
% Zeros of nth-order ordinary Bessel polynomial ... are the same as for modified
% Bessel function of second kind K_v:
% exp(1/x) * K_{n+0.5}(1/x) = ...
% Hence, we define the target function and its first derivative as
```
Where you have to fill the `...` part. Is it also second kind and then `Y`?

**Comment by hagenw** [comment]
I added a numerical test case as well for the order 1, 10, and 100 comparing to the results from the python code. This is not perfect as the outcome depends on the ordering of the zeros in the resulting vector, which can differ for different calculation methods, but I thought it is still better than nothing.

File: SFS_general/sphbesselh_zeros.m [hunk 2]
```diff
@@ -53,74 +66,114 @@
...
-% The following code was used to calculate the zeros with the Multiprecission
-% Toolbox. The results are stored at
-% https://github.com/sfstoolbox/data/tree/master/sphbesselh_zeros
-B = mp(zeros(1,order+2));
-A = B;
-for n=mp(0:order)
-    B(n+1) = factorial(2*order-n)/(factorial(order-n)*factorial(n)*2^(order-n));
+
+function x = newton(f,fp,x0,TOL,MAXITER)
+    % Newton-Rapheson method for approximation of a single root
```
Comment (hagenw) [inline SFS_general/sphbesselh_zeros.m]: Where comes this implementation from?
When I'm looking at [scipy's newton implementation](https://github.com/scipy/scipy/blob/07dd175d5897ebc4dcf8e7f42eb4f2c34f005fe3/scipy/optimize/zeros.py#L65-L184) it looks a lot more complicated.

**Comment by hagenw** [comment]
I also did a NFC-HOA calculation in order to compare with Hahn and Spors (2017) Fig. 8:
```Matlab
conf = SFS_config;
conf.nfchoa.order = 30;
sound_field_imp_nfchoa([-2 2],[-2 2],0,[1.5 1.5 0],'ps',1/conf.c,conf)
```
nfchoa_n30
```Matlab
conf.nfchoa.order = 120;
sound_field_imp_nfchoa([-2 2],[-2 2],0,[1.5 1.5 0],'ps',1/conf.c,conf)
```
nfchoa_n120

It looks very similar, but we have the different of that prominent circular contribution which is not part of the plot in the paper. If I remember correctly we had that discussion already somewhere else, hadn't we?
Is this due to the inclusion/exclusion of the lowest order?

File: SFS_general/sphbesselh_zeros.m (hunk 2 above)
Comment (fietew) [inline SFS_general/sphbesselh_zeros.m]: The scipy implementation does handle many special cases and has some extensions. Here, the most simple version of the Newton-Rapheson method is implemented.

**Comment by fietew** [comment]
> For the test function, could you maybe send me a screen shot of the Matlab result, that I can try to get the same with Octave.

I am pretty sure, that the coloring is not relevant, but I will explicitly define the colors in the script and send you a screen shot.

> Is this due to the inclusion/exclusion of the lowest order?
What are you referring to? What is the lowest order? What is included/excluded?

**Comment by fietew** [comment]
The plot under MATLAB R2015a:
zeros

File: SFS_general/sphbesselh_zeros.m (hunk 2 above)
Comment (hagenw) [inline SFS_general/sphbesselh_zeros.m]: Now, I found also in the scipy version. Then maybe the easiest thing is to add not a link to the function header, but simply go with the current help text for `newton()`.

File: SFS_general/sphbesselh_zeros.m (hunk 1 above)
Comment (hagenw) [inline SFS_general/sphbesselh_zeros.m]: Could you help to fill the `...` please.

File: SFS_general/sphbesselh_zeros.m (hunk 2 above)
Comment (fietew) [inline SFS_general/sphbesselh_zeros.m]: I can't follow you. What do you want me to do?

**Comment by hagenw** [comment]
The colors in the plot work now also under Octave, but the legend was still far to big. I added a commit that replaces the legend with an updated title, indicating the shown orders for every plot.

**Comment by hagenw** [comment]
Coming back to the computed figure and the comparison to Hahn and Spors (2017) Fig. 8. Forget what I have set about the order. But do you have an idea why we have the obvious difference between those plots (of course besides the difference in amplitude)?

File: SFS_general/sphbesselh_zeros.m (hunk 2 above)
Comment (hagenw) [inline SFS_general/sphbesselh_zeros.m]: Nothing :)
Just a confirmation that it is fine to not add a reference link to the scipy code for the Newton method.

**Comment by fietew** [comment]
Are you using normalisation for the plot? Also the range of the colormap in  Hahn and Spors (2017) Fig. 8. is [-2,2].

**Comment by hagenw** [comment]
I'm using normalisation, otherwise there is no energy in the plot. But you can do the following to get a very similar energy as in Hahn and Spors (2017) Fig. 8:
```Matlab
conf = SFS_config;
conf.nfchoa.order = 30;
conf.plot.caxis = [-0.3 0.3];
sound_field_imp_nfchoa([-2 2],[-2 2],0,[1.5 1.5 0],'ps',1/conf.c,conf)
```
will result in
nfchoa

This looks similar, but still we have a more pronounced visible circle as indicated by the arrow in the figure.

**Comment by fietew** [comment]
Was #56 the discussion you meant?

**Comment by fietew** [comment]
Is it possible, that you have chosen a different number of loudspeakers? In the paper, 60 has been chosen.
For 64 loudspeaker, one loudspeaker is exactly at 45 degrees, which looks like the virtual point source direction. For 60 loudspeakers, 45 degrees lies between two loudspeakers. Maybe this is the reason.

**Comment by hagenw** [comment]
Ah, brilliant, that's it:
```Matlab
conf = SFS_config;
conf.nfchoa.order = 30;
conf.secondary_sources.number = 60;
conf.plot.caxis = [-0.3 0.3];
sound_field_imp_nfchoa([-2 2],[-2 2],0,[1.5 1.5 0],'ps',1/conf.c,conf)
```
and we will get the same result as in the paper.
nfchoa2

I would say this is then nearly ready. There is only [this discussion on the Bessel polynomial](https://github.com/sfstoolbox/sfs-matlab/pull/152#pullrequestreview-48585974) left, @narahahn should give his go, and after a small commit clean up we could merge.

File: SFS_general/sphbesselh_zeros.m (hunk 1 above)
Comment (fietew) [inline SFS_general/sphbesselh_zeros.m]: According to https://en.wikipedia.org/wiki/Bessel_polynomials#Definition_in_terms_of_Bessel_functions
the nth-order ordinary bessel polynomial is denoted by ``y_n``, which is also the spherical Neumann function in our nomenclature. I would suggest:

```
% Zeros of nth-order ordinary Bessel polynomial y_n are the same as for 
% the exponentially scaled modified Bessel function of second kind K_v:
% \sqrt{2pi/x} * exp(1/x) * K_{n+0.5}(1/x) = y_n
% Hence, we define the target function and its first derivative as
```

**Comment by hagenw** [comment]
I made another test and tried to improve [Fig. 3.14 from my thesis](https://github.com/hagenw/phd-thesis/tree/master/03_sound_field_errors_and_their_perceptual_relevance/fig3_14) and it worked quite well.

Original NFC-HOA part of the figure:
nfchoa_old

New NFC-HOA part of the figure as calculated with this branch:
nfchoa_new

@fietew: From my side, I have no further change requests for this branch and would like to start cleaning the commit history up, if you agree.

**Comment by fietew** [comment]
I agree.

**Pull request merged**